sympy/core/operations.py

```diff
@@ -504,3 +505,113 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher:
+    """
+    Handler dispatcher for associative operators
+
+    .. notes::
+       This approach is experimental, and can be replaced or deleted in the future.
+       See https://github.com/sympy/sympy/pull/19463.
+
+    Explanation
+    ===========
+
+    If a new class is meant to return the result other than `Add` or `Mul` when operated,
+    it should have its handler defined as class method. Also, if the arguments have the same
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if isinstance(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.name = name
+        self.fallback = fallback
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self.dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
```

Comment:
Your latest change makes this no longer true. The new description seems to be
```suggestion
        Register the priority between two types. Registered function must
        return an arbitrary and irrelevant object that has the type which has the handler that is to be used.
```

I think I preferred the previous design.

Reply:
Previous design before 731ad8f was inconsistent so I resolved it first. I'll implement your suggestion tomorrow.

Reply:
Inconsistent with respect to what?